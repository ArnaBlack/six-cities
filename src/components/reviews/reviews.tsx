import * as React from 'react';
import {connect} from 'react-redux';

import ReviewList from '../review-list/review-list';
import ReviewForm from '../review-form/review-form';

import DataOperation from '../../store/data/operation/operation';

import {
  getLoadingState,
  getReviews,
} from '../../store/data/selectors';

import {Comment} from '../../types';

interface Props {
  isLoading: boolean,
  placeId: number,
  reviews: Comment[],
  loadReviews: (id: number) => void,
}

class Reviews extends React.PureComponent<Props, null> {
  render() {
    const {
      isLoading,
      reviews,
    } = this.props;

    return isLoading ? null : <section className="property__reviews reviews">
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
      <ReviewList />
      <ReviewForm />
    </section>;
  }

  componentDidMount() {
    const {
      loadReviews,
      placeId,
    } = this.props;

    loadReviews(placeId);
  }
}

const mapStateToProps = (state, props) => ({
  ...props,
  isLoading: getLoadingState(state),
  reviews: getReviews(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadReviews: (id) => dispatch(DataOperation.loadReviews(id))
});

export {Reviews};
export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
