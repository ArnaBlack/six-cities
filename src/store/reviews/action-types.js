export const LOAD_REVIEWS = `LOAD_REVIEWS`;
export const SEND_REVIEW = `SEND_REVIEW`;
export const SEND_REVIEW_SUCCESS = `SEND_REVIEW_SUCCESS`;
export const SEND_REVIEW_ERROR = `SEND_REVIEW_ERROR`;
