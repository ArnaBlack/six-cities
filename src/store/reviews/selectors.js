import NameSpace from '../name-space';
import {MAX_REVIEWS} from '../../constants';

export const getLoadingState = (state) => state[NameSpace.REVIEWS].isLoading;
export const getSendingState = (state) => state[NameSpace.REVIEWS].isSending;
export const getSendingError = (state) => state[NameSpace.REVIEWS].sendingError;
export const getReviews = (state) => state[NameSpace.REVIEWS].reviews
  .sort((a, b) => new Date(b.date) - new Date(a.date))
  .slice(0, MAX_REVIEWS);
