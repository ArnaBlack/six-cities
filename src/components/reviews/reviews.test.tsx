import * as React from 'react';
import {
  configure,
  shallow,
} from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import {Reviews} from './reviews';

configure({adapter: new Adapter()});

const mock = {
  isLoading: false,
  isAuthorizationRequired: false,
  placeId: 1,
  reviews: [
    {
      id: 1,
      user: {
        avatarUrl: ``,
        id: 1,
        isPro: false,
        name: ``,
      },
      rating: 4,
      comment: ``,
      date: ``,
    },
  ],
};

it(`Reviews correctly renders`, () => {
  const {
    isLoading,
    isAuthorizationRequired,
    placeId,
    reviews,
  } = mock;
  const loadReviews = jest.fn();
  const tree = shallow(<Reviews
    isLoading={isLoading}
    isAuthorizationRequired={isAuthorizationRequired}
    placeId={placeId}
    reviews={reviews}
    loadReviews={loadReviews}
  />);

  expect(tree).toMatchSnapshot();
});
