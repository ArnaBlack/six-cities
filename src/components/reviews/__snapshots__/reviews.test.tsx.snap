// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews correctly renders 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      1
    </span>
  </h2>
  <ConnectFunction />
  <ReviewForm />
</section>
`;
